hyperspy/model.py

```diff
@@ -1133,19 +1132,37 @@ def fit(self, fitter="leastsq", method='ls', grad=False,
                     # This replicates the original "leastsq"
                     # behaviour in earlier versions of HyperSpy
                     # using the Levenberg-Marquardt algorithm
-                    output = \
-                        leastsq(self._errfunc, self.p0[:], Dfun=jacobian,
-                                col_deriv=1, args=args, full_output=True,
-                                **kwargs)
-                    self.p0, pcov = output[0:2]
-
-                signal_len = sum([axis.size
-                                  for axis in self.axes_manager.signal_axes])
-                if (signal_len > len(self.p0)) and pcov is not None:
-                    pcov *= ((self._errfunc(self.p0, *args) ** 2).sum() /
-                             (len(args[0]) - len(self.p0)))
-
-                    self.p_std = np.sqrt(np.diag(pcov))
+                    output = leastsq(self._errfunc, self.p0[:], Dfun=jacobian,
+                                     col_deriv=1, args=args, full_output=True,
+                                     **kwargs)
+                    self.p0, pcov, infodict, errmsg, ier = output
+                    ysize = len(infodict['fvec'])
+                    cost = np.sum(infodict['fvec'] ** 2)
+
+                warn_cov = False
+
+                if pcov is None:
+                    # Indeterminate covariance
+                    pcov = np.zeros((len(self.p0), len(self.p0)), dtype=float)
+                    pcov.fill(np.inf)
+                    warn_cov = True
+                elif pcov.min() < 0:
+                    # Usually indicative of numerical overflow
+                    # (covariance should be positive)
+                    pcov.fill(np.inf)
+                    warn_cov = True
+                elif (ysize > self.p0.size):
+                    pcov *= cost / (ysize - self.p0.size)
+                else:
+                    pcov.fill(np.inf)
+                    warn_cov = True
+
+                if warn_cov:
+                    _logger.warning("Covariance of the parameters could not be estimated")
```

Comment:
Even if this warning message is definitely better than the previous ones, would it be possible to be more specific for each of the cases above? I have no idea if this is possible/sensible or not.
Also what does it implies will not be clear for many users: is it worth adding that the error of the parameters will be `inf`? And that the model doesn't fit well the data?

Reply:
Yes I totally agree - I will be more specific with the warning message raised - doing that now.

Reply:
@ericpre  that's updated now.

Reply:
> Even if this warning message is definitely better than the previous ones, would it be possible to be more specific for each of the cases above? I have no idea if this is possible/sensible or not.

I've not created specific warnings since they all amount to the same thing - for example, `pcov.min() < 0` usually happens due to an ill-conditioned matrix inversion within the scipy MINPACK `leastsq` function leading to overflow, but this is effectively the same reason as `pcov = None` - the Hessian is singular, so can't be inverted.